From b41e5db04c3c828869c04f386de5c09cdebeeac3 Mon Sep 17 00:00:00 2001
From: Reto Schneider <code@reto-schneider.ch>
Date: Thu, 23 Aug 2018 02:39:07 +0200
Subject: [PATCH] Do url-encoding of file names using the curl backend

Warning: This hack requires the usage of the curl backend.

---
 libopkg/opkg_download.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/libopkg/opkg_download.c b/libopkg/opkg_download.c
index 480ae02..135f68c 100644
--- a/libopkg/opkg_download.c
+++ b/libopkg/opkg_download.c
@@ -19,6 +19,7 @@
 
 #include "config.h"
 
+#include <curl/curl.h>
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
@@ -227,6 +228,8 @@ int opkg_download(const char *src, const char *dest_file_name,
 static char *get_pkg_url(pkg_t * pkg)
 {
     char *url;
+    const char *filename_encoded;
+    CURL *curl;
 
     if (pkg->src == NULL) {
         opkg_msg(ERROR,
@@ -240,7 +243,21 @@ static char *get_pkg_url(pkg_t * pkg)
         return NULL;
     }
 
-    sprintf_alloc(&url, "%s/%s", pkg->src->value, pkg->filename);
+    if (!str_starts_with(pkg->src->value, "file:")) {
+        curl = curl_easy_init();
+        if (!curl)
+            return NULL;
+
+        filename_encoded = curl_easy_escape(curl, pkg->filename, 0);
+
+        sprintf_alloc(&url, "%s/%s", pkg->src->value, filename_encoded);
+
+        curl_free(filename_encoded);
+        curl_easy_cleanup(curl);
+    } else {
+        sprintf_alloc(&url, "%s/%s", pkg->src->value, pkg->filename);
+    }
+
     return url;
 }
 
